netty-server/src/main/scala/org/http4s/netty/Http4sNettyHandler.scala

```diff
@@ -91,37 +101,49 @@ sealed abstract class MikuHandler[F[_]](
                 p.complete(Failure(error)); ()
               }
 
-            case Right(httpResponse) =>
+            case Right(r) =>
               IO {
-                p.complete(Success(httpResponse)); ()
+                p.complete(Success(r)); ()
               }
 
           }
           .unsafeRunSync()
-        val futureResponse: Future[HttpResponse] = p.future
 
         //This attaches all writes sequentially using
         //LastResponseSent as a queue. `trampoline` ensures we do not
         //CTX switch the writes.
         lastResponseSent = lastResponseSent.flatMap[Unit] { _ =>
-          futureResponse
-            .map[Unit] { response =>
-              if (requestsInFlight.decrementAndGet() == 0) {
-                // Since we've now gone down to zero, we need to issue a
-                // read, in case we ignored an earlier read complete
-                ctx.read()
-              }
-              ctx.writeAndFlush(response); ()
+          p.future
+            .map[Unit] {
+              case (response, cleanup) =>
+                if (requestsInFlight.decrementAndGet() == 0) {
+                  // Since we've now gone down to zero, we need to issue a
+                  // read, in case we ignored an earlier read complete
+                  ctx.read()
+                }
+                ctx
+                  .writeAndFlush(response)
+                  .addListener(new ChannelFutureListener {
+                    def operationComplete(future: ChannelFuture): Unit =
+                      //Run cleanup action on the user-supplied pool.
+                      //We might tweak this later on to run on the same or in a cleanup pool
+                      //Benching this would be interesting
+                      ec.execute(new Runnable {
+                        def run(): Unit =
+                          F.runAsync(cleanup(future.channel()))(_ => IO.unit).unsafeRunSync()
```

Comment:
cats-effect has `Logger.reportFailure` for this.  We should add that to our own internals.  This `_ => IO.unit` swallows errors.

Reply:
Will add.

Reply:
Hmm actually, reconsidering here:

The cleanup action is either just `unit` or `drainBody`, neither of which will throw an exception. 

Save a VM fatal error, `drainBody` doesn't fail/throw. I'd like to reconsider this one.

Reply:
I think you're right, but I also think I've wasted a lot of my life thinking I was right and finding the murder weapon was hidden in an empty exception handler.  I guess I reflexively do this, if only to make it look even more jarring when people don't.

I won't die on this hill.

Reply:
If it's any consolation whatsoever:

This part I spent a _lot_ of time on. In particular, because that cleanup action drains the buffers out of the reactive stream publisher if it wasn't used. This would be a huge memleak if it turned out to be an issue.

I spent hours debugging this and ensuring resources were released properly. I silently discard errors here because I came to the conclusion after a lot of moving around, spamming requests and observing output:
- If even one element of the stream was consumed this no-ops.
- This is primarily concerned with resource release and it only fires _after_ the channel write is complete. If it ever errors out (extremely low chance), it still means the stream was consumed somewhere else and thus released properly. It can only error if you consumed the stream and closed it and then tried to consume it again, in which case the upstream publisher is already done releasing resources.

I really hope the confidence that I gained by spending a long time on this doesn't come back to bite me, but I'm fairly confident this is right.